### PR TITLE
Update test-runner api

### PR DESCRIPTION
**Pull request by seunlanlege**
This refactors the test-runner api.

polkadot companion: paritytech/polkadot#2609

### CODE REVIEW

File: test-utils/test-runner/Cargo.toml [hunk 1]
```diff
@@ -47,6 +51,9 @@ env_logger = "0.7.1"
 log = "0.4.8"
 futures = { package = "futures", version = "0.3", features = ["compat"] }
 tokio = { version = "0.2", features = ["full"] }
-
+parking_lot = "0.11.1"
 # Calling RPC
 jsonrpc-core = "15.1"
+jsonrpc-core-client = "15.1"
+jsonrpc-derive = "15.1"
```
Comment (ordian) [inline test-utils/test-runner/Cargo.toml:58]: `unused_extern_crates`?

File: test-utils/test-runner/src/utils.rs [hunk 2]
```diff
@@ -150,3 +124,15 @@ pub fn default_config(task_executor: TaskExecutor, mut chain_spec: Box<dyn Chain
 		transaction_storage: TransactionStorageMode::BlockBody,
 	}
 }
+
+/// Produce a task executor given a handle to a tokio runtime
+pub fn task_executor(handle: Handle) -> TaskExecutor {
+	let task_executor = move |fut, task_type| match task_type {
+		TaskType::Async => handle.spawn(fut).map(drop),
+		TaskType::Blocking => handle
+			.spawn_blocking(move || futures::executor::block_on(fut))
```
Comment (ordian) [inline test-utils/test-runner/src/utils.rs:133]: is that a common thing to run one executor inside of the other? seems weird to me

File: test-utils/test-runner/src/node.rs [hunk 3]
```diff
@@ -353,30 +209,29 @@ impl<T: ChainInfo> Node<T> {
 
 	/// Revert all blocks added since creation of the node.
 	pub fn clean(&self) {
-		// if a db path was specified, revert all blocks we've added
-		if let Some(_) = std::env::var("DB_BASE_PATH").ok() {
+		// if the initial block number wasn't 0, then revert blocks
+		if self.initial_block_number != 0u32.into() {
```
Comment (ordian) [inline test-utils/test-runner/src/node.rs:213]: why?

File: test-utils/test-runner/src/node.rs [hunk 4]
```diff
@@ -353,30 +209,29 @@ impl<T: ChainInfo> Node<T> {
...
-	/// Performs a runtime upgrade given a wasm blob.
-	pub fn upgrade_runtime(&mut self, wasm: Vec<u8>)
-		where
-			<T::Runtime as frame_system::Config>::Call: From<frame_system::Call<T::Runtime>>
-	{
-		let call = frame_system::Call::set_code(wasm);
-		T::dispatch_with_root(call.into(), self);
+	/// so you've decided to run the test runner as a binary, use this to shutdown gracefully.
+	pub async fn until_shutdown(&self) {
+		let manager = self.task_manager.lock().take();
```
Comment (ordian) [inline test-utils/test-runner/src/node.rs:221]: using `parking_lot::Mutex` inside of an async function doesn't look like a good idea to me

**Review by ordian — COMMENTED** [review]
I'm not familiar with code, so would appreciate if someone else would take a look.

File: test-utils/test-runner/src/node.rs (hunk 3 above)
Comment (seunlanlege) [reply test-utils/test-runner/src/node.rs:213]: One of the use cases of the test runner is running it over live chain data, in this case it's `initial_block_number` is non-zero so we want to revert any blocks we've added, so we can re-use the live chain data.

File: test-utils/test-runner/src/node.rs (hunk 4 above)
Comment (seunlanlege) [reply test-utils/test-runner/src/node.rs:221]: how about `tokio::sync::Mutex`?

File: test-utils/test-runner/src/utils.rs (hunk 2 above)
Comment (seunlanlege) [reply test-utils/test-runner/src/utils.rs:133]: yeah it's literally how sc_cli does it

https://github.com/paritytech/substrate/blob/d3db3c1d1011983ba7565e342910a6d2095625c8/client/cli/src/runner.rs#L129-L136

File: test-utils/test-runner/src/node.rs (hunk 3 above)
Comment (ordian) [reply test-utils/test-runner/src/node.rs:213]: I understand that, but why special-casing initial_block_number == 0? what if started from 0, but still want to revert all blocks?

File: test-utils/test-runner/src/node.rs (hunk 4 above)
Comment (ordian) [reply test-utils/test-runner/src/node.rs:221]: But is it really needed? Is it used only to allow interior mutability? If so, can you change the signature of the function to `&mut self`?

Comment (seunlanlege) [reply test-utils/test-runner/src/node.rs:221]: > But is it really needed? 

ahh, you're right, this really shouldn't be called from multiple threads.

File: test-utils/test-runner/Cargo.toml [hunk 5]
```diff
@@ -43,10 +46,10 @@ sp-runtime-interface = { version = "3.0.0", path = "../../primitives/runtime-int
 # pallets
 frame-system = { version = "3.0.0", path = "../../frame/system" }
 
-env_logger = "0.7.1"
 log = "0.4.8"
 futures = { package = "futures", version = "0.3", features = ["compat"] }
-tokio = { version = "0.2", features = ["full"] }
-
+tokio = { version = "0.2", features = ["signal"] }
+parking_lot = "0.11.1"
```
Comment (ordian) [inline test-utils/test-runner/Cargo.toml:52]:
```suggestion
```

File: test-utils/test-runner/src/node.rs (hunk 3 above)
Comment (seunlanlege) [reply test-utils/test-runner/src/node.rs:212]: you know what you're right about this one too, it's a test environment, we should always revert.

**Comment by seunlanlege** [comment]
@ordian think this ok now?

**Comment by ordian** [comment]
> @ordian think this ok now?

I'm not familiar with this code enough to be confident to give an approval, so would appreciate if someone else would do it.

**Review by tomusdrw — APPROVED** [review]
Not sure what the context is, but I'm assuming it's exposing the `async/await` functions instead of having a blocking API?

I also don't know exactly where/if this code is used, but the changes looks sane to me.

**Comment by seunlanlege** [comment]
ui tests won't let me great :(

**Review by ordian — APPROVED** [review]

**Pull request merged**